client/internal/providerqueryer/queryer_test.go

```diff
@@ -0,0 +1,466 @@
+						{numProvsFound: 0},
+					},
+				},
+				{
+					advance: findProviderTimeout,
+					expStates: []reqState{
+						{numProvsFound: 0},
+					},
+				},
+			},
+			expQueries: 1,
+		},
+		{
+			name:         "query completes even if the context is already canceled",
+			numProviders: 1,
+			numKeys:      1,
+			provDelay:    1 * time.Millisecond,
+			cancelCtx:    true,
+			reqs: []req{
+				{keyIdx: 0},
+			},
+			epochs: []epoch{
+				{
+					advance: 0,
+					expStates: []reqState{
+						{numProvsFound: 0},
+					},
+				},
+			},
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			peers := testutil.GeneratePeers(c.numProviders)
+			mockClock := clock.NewMock()
+			fpn := &fakeProviderNetwork{
+				clock:      mockClock,
+				peersFound: peers,
+				delay:      c.provDelay,
+			}
+
+			if c.connErrs {
+				fpn.connectError = errors.New("unable to connect")
+			}
+
+			ctx := context.Background()
+			ctx, cancel := mockClock.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+
+			if c.cancelCtx {
+				cancel()
+			}
+
+			queryer := New(fpn)
+			queryer.clock = mockClock
+
+			if c.maxConcurrentReqs != 0 {
+				queryer.sem = semaphore.NewWeighted(c.maxConcurrentReqs)
+			}
+
+			keys := testutil.GenerateCids(c.numKeys)
+
+			var bufs []*reqBuf
+			for _, req := range c.reqs {
+				reqCtx := ctx
+				var reqCancel func()
+				if req.timeout > 0 {
+					reqCtx, reqCancel = mockClock.WithTimeout(reqCtx, req.timeout)
+					defer reqCancel()
+				}
+				k := keys[req.keyIdx]
+				buf := &reqBuf{key: k}
+				buf.run(func() <-chan peer.ID {
+					return queryer.FindProvidersAsync(reqCtx, k)
+				})
+				bufs = append(bufs, buf)
+				// fire clock events
+				mockClock.Add(0)
+			}
+
+			for epochIdx, epoch := range c.epochs {
+				mockClock.Add(epoch.advance)
+				for reqIdx, reqState := range epoch.expStates {
+					buf := bufs[reqIdx]
+					assert.Equal(t, reqState.numProvsFound, len(buf.getPeers()), "unexpected number of results in epoch=%d for req=%d", epochIdx, reqIdx)
+				}
+			}
+
+			for _, buf := range bufs {
+				buf.wait()
+			}
+
+			fpn.queriesMadeMutex.Lock()
+			defer fpn.queriesMadeMutex.Unlock()
+			assert.Equal(t, c.expQueries, fpn.queriesMade, "didn't make expected number of downstream queries")
+		})
+	}
+
+}
+
+func TestQueryer_Fuzz(t *testing.T) {
+	// Fuzz the queryer to check for races, deadlocks, etc.
+	//
+	// We don't use Go's fuzz testing framework because it constantly craps out
+	// due to an unconfigurable per-test timeout. Once that's fixed we can probably
+	// switch to it.
+
+	deadlock.Opts.DeadlockTimeout = 1 * time.Second
+
+	// switch to using instrumented locks to detect deadlocks
+	origNewMutex := newMutex
+	origNewRWMutex := newRWMutex
+	newMutex = func() sync.Locker { return &deadlock.Mutex{} }
+	newRWMutex = func() rwLocker { return &deadlock.RWMutex{} }
+	defer func() { newMutex = origNewMutex }()
+	defer func() { newRWMutex = origNewRWMutex }()
```

Comment:
I don't think you need this, if the code deadlock on a lock this is gonna be caught by the:
```go
			case <-ctx.Done():
				t.Fatalf("timeout")
```
Clause anyway

Reply:
it's useful because it tells you why the deadlock occurred (what last grabbed the lock, which goroutines are stuck on the lock, etc.)